### PR TITLE
[AIRFLOW-5704] Improve Kind Kubernetes scripts for local testing

### PR DESCRIPTION
**Pull request by potiuk**
* Fixed problem that Kubernetes tests were testing latest master
  rather than what came from the local sources.
* Kind (Kubernetes in Dcocker) is run in the same Docker as Breeze env
* Moved Kubernetes scripts to 'in_container' dir where they belong now
* Kubernetes cluster is reused until it is stopped
* Kubernetes image is build from image in docker already + mounted sources
* Kubectl version name is corrected in the Dockerfile
* KUBERNETES_VERSION can now be used to select Kubernetes version
* Running kubernetes scripts is now easy in Breeze
* We can start/recreate/stop cluster using  --<ACTION>-kind-cluster
* Instructions on how to run Kubernetes tests are updated
* The old "bare" environment is replaced by --no-deps switch

---
Issue link: [AIRFLOW-5704](https://issues.apache.org/jira/browse/AIRFLOW-5704/)

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by potiuk** [comment]
@ashb @dimberman @gerardo -> I merged all the smaller PRs separated out from the Kind change and I am working on making the tests work for this one. While trying to fix and tests the kubernetes tests I noticed how difficult it was to iterate and re-test Kubernetes tests, so this one is now optimising even more:

1) Only the basic image for CI is built outside and sent to the docker. It is sent via untarred folder rather than as whole file. This makes it much nicer for the slow Mac Host->Docker volume mounting because the untarred image save contains separate .tars for all Docker image layers. I sync the layer dir via rsync based on checksum - this way I make sure than whenever the CI image is rebuilt, only the last rebuilt layers are changed. Then when the image is imported to DinD only the last layers are updated. this way while the environment is up and running and images are loaded to DiND and cluster, only incremental changes will be loaded.

2) Then I came back to building the kubernetes image inside breeze container. I will speed it up more, but it is actually much better for iterating when you are inside the breeze container. In order to reload last image to the cluster you will just need to run "rebuild_ariflow_image.sh" inside the container and it will (incrementally of course) rebuild and sent the new image layers to DiND and then to cluster.

I am still working on fixing the tests but I think it is much more developer-friendly than before.

**Comment by davlum** [comment]
I'm getting an error `/opt/airflow/scripts/ci/in_container/kubernetes/app/deploy_app.sh: line 78: AIRFLOW_CI_KUBERNETES_IMAGE: unbound variable` when running `./breeze -K`. 

If that's a required argument it should fail up front.

File: BREEZE.rst [hunk 1]
```diff
@@ -528,6 +509,54 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+
+Building Airflow Images
+-----------------------
+
+This is done using ``scripts/ci/in_container/kubernetes/setup_kubernetes.sh`` script:
+
+1. Latest ``apache/airflow:master-pythonX.Y-ci`` images are pulled from DockerHub to the 'docker' container.
+2. The image is rebuild to take into account the latest sources mounted to Breeze environment.
+3. Scripts required for running the image as Kubernetes cluster are added and
+   ``apache/airflow:master-pythonX.Y-ci-kubernetes``
```
Comment (mik-laj) [inline BREEZE.rst]:
```diff
-   ``apache/airflow:master-pythonX.Y-ci-kubernetes``
+   ``apache/airflow:master-pythonX.Y-ci-kubernetes`` image.
```

File: BREEZE.rst [hunk 2]
```diff
@@ -528,6 +509,54 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+=======================================
+
+In order to run Kubernetes in Breeze you can start Breeze with ``--start-kind-cluster`` switch. This will
+automatically create a Kind Kubernetes cluster in a ``docker`` (docker-in-docker) container started as part
+of Breeze environment. Setting up the Kubernetes cluster takes some time so the cluster continues running
+until the environment is stopped with ``--stop-environment`` switch or until ``--recreate-kind-cluster``
+switch is used additionally to ``--start-kind-cluster``.
+
+The Kubernetes cluster is automatically initialized and starts Airflow webserver and scheduler
+using latest sources. This is done in two steps. First Airflow images are build in the ``docker`` image,
```
Comment (mik-laj) [inline BREEZE.rst]:
```
using the latest sources. This is done in two steps. First Airflow images are built in the ``docker`` image,
```
I'm not sure.

File: BREEZE.rst [hunk 3]
```diff
@@ -528,6 +509,54 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+then Airflow applications (``scheduler``/``webserver``) are prepared and deployed at the kind cluster using
+the images prepared.
+
+Rebuilding images should be faster every next time we enter the environment but it is attempted every time
+you restart breeze environment. Similarly applications are re-deployed every time the breeze environment
```
Comment (mik-laj) [inline BREEZE.rst]:
```diff
-you restart breeze environment. Similarly applications are re-deployed every time the breeze environment
+you restart breeze environment. Similarly, applications are re-deployed every time the breeze environment

```

File: BREEZE.rst [hunk 4]
```diff
@@ -646,22 +675,29 @@ This is the current syntax for  `./breeze <./breeze>`_:
...
-
   -B, --backend <BACKEND>
           Backend to use for tests - it determines which database is used.
           One of [ sqlite mysql postgres ]. Default: sqlite
 
-  -K, --kubernetes-version <KUBERNETES_VERSION>
-          Kubernetes version - only used in case of 'kubernetes' environment.
-          One of [ v1.13.0 ]. Default: v1.13.0
+  -K, --start-kind-cluster
+          Starts kind kubernetes cluster after entering the environment. The cluster is started using
```
Comment (mik-laj) [inline BREEZE.rst]:
```diff
-          Starts kind kubernetes cluster after entering the environment. The cluster is started using
+          Starts kind Kubernetes cluster after entering the environment. The cluster is started using

```

File: BREEZE.rst [hunk 5]
```diff
@@ -646,22 +675,29 @@ This is the current syntax for  `./breeze <./breeze>`_:
...
+          Kubernetes Mode selected and Kubernetes version specifed via --kubernetes-mode and
+          --kubernetes-version flags.
+
+  -Z, --recreate-kind-cluster
+          Recreates kind kubernetes cluster if one has already been created. By default if you do not stop
```
Comment (mik-laj) [inline BREEZE.rst]:
```diff
-          Recreates kind kubernetes cluster if one has already been created. By default if you do not stop
+          Recreates kind Kubernetes cluster if one has already been created. By default, if you do not stop

```

File: BREEZE.rst [hunk 6]
```diff
@@ -646,22 +675,29 @@ This is the current syntax for  `./breeze <./breeze>`_:
...
+          environment, the kubernetes cluster created for testing is continuously running and when
```
Comment (mik-laj) [inline BREEZE.rst]: kubernetes => Kubernetes

**Review by mik-laj — APPROVED** [review]

File: BREEZE.rst (hunk 2 above)
Comment (potiuk) [reply BREEZE.rst]: Clarified.

File: BREEZE.rst (hunk 1 above)
Comment (potiuk) [reply BREEZE.rst]: clarified

File: BREEZE.rst (hunk 2 above)
Comment (mik-laj) [reply BREEZE.rst]:
```diff
-using latest sources. This is done in two steps. First Airflow images are build in the ``docker`` image,
+using the latest sources. This is done in two steps. First Airflow images are built in the ``docker`` image,
```
Small typo. latest => the latest, build=> built

**Comment by potiuk** [comment]
> I'm getting an error `/opt/airflow/scripts/ci/in_container/kubernetes/app/deploy_app.sh: line 78: AIRFLOW_CI_KUBERNETES_IMAGE: unbound variable` when running `./breeze -K`.
> 
> If that's a required argument it should fail up front.

That was a bug. I am working on fixing it @davlum

**Comment by potiuk** [comment]
Hey @gerardo @dimberman @kaxil @ashb @mik-laj @nuclearpinguin 

I think you all might be interested with this change to see/review/understand how kind implementation evolves and how Breeze will let people to work with Kubernetes Executor in development mode. Kind seems to be really nice for both CI and local development.

I have finally implemented the "best" approach with Kind I think. One that is much faster and seems to work really well for both - CI (much faster than previous approach with separate docker container) - it also helps with local development with kind cluster . Now I am testing all the ins/outs of it but it seems that it is really nice for local development as well (with regards to resources/time pulling images/using latest sources etc.). Together with the follow-up change of separating integration and slimming down Breeze: #7091 - it becomes really useful for local development.

Some of the notable properties:

- I switched to the latest kind 6.1 (it does not require separate exporting of the ".kube folder - it is done automatically) 

- I pass the docker socket from host to the container for "ENABLE_KIND_CLUSTER" build - this way I can use the same docker image that airflow-testing is run on - all the Kind containers are run in that docker image (and visible from the host via docker ps). This means a) no separate docker container is needed b) no need to pull image from apache/airflow - it's already there! I am checking if container is already created and skip creating it if so. You can also force cluster recreation by passing --recreate-kind cluster flag or stop 

- this also means that if I exit the Breeze/airflow-testing container - the cluster is still running there and I do not recreate the cluster when I re-enter the container. 

- However I still run "rebuild_airflow_image.sh" when entering the container. This way I am 100% sure I use latest sources (both locally and in CI). Plus I can later change the sources and run "rebuild_airflow_image.sh" without leaving the airflow-testing container! What's nice about it - it's rather fast as it only rebuilds latest layers. It allows to iterate semi-quickly while working on Kubernetes :). Still loading image to kind takes quite some time (6 minutes for me) so it is not perfect. I will improve it when I have prod image - it should be much faster to load rather than the CI one

- last but not least - the .kube folder is mounted from the HOST (${AIRFLOW_SOURCES}/.kube) so you should be able to interact with the cluster from the host as well (You just need to update hostname to localhost). I might want to document it or maybe even automatically copy the configuration to include the configuration that will support kubectl/connecting to cluster-run webserver from the localhost directly. All ports are exposed already so it should be rather easy.

**Comment by kaxil** [comment]
> @ashb @dimberman @gerardo -> I merged all the smaller PRs separated out from the Kind change and I am working on making the tests work for this one. While trying to fix and tests the kubernetes tests I noticed how difficult it was to iterate and re-test Kubernetes tests, so this one is now optimising even more:
> 
> 1. Only the basic image for CI is built outside and sent to the docker. It is sent via untarred folder rather than as whole file. This makes it much nicer for the slow Mac Host->Docker volume mounting because the untarred image save contains separate .tars for all Docker image layers. I sync the layer dir via rsync based on checksum - this way I make sure than whenever the CI image is rebuilt, only the last rebuilt layers are changed. Then when the image is imported to DinD only the last layers are updated. this way while the environment is up and running and images are loaded to DiND and cluster, only incremental changes will be loaded.
> 2. Then I came back to building the kubernetes image inside breeze container. I will speed it up more, but it is actually much better for iterating when you are inside the breeze container. In order to reload last image to the cluster you will just need to run "rebuild_ariflow_image.sh" inside the container and it will (incrementally of course) rebuild and sent the new image layers to DiND and then to cluster.
> 
> I am still working on fixing the tests but I think it is much more developer-friendly than before.

Nice work @potiuk

**Comment by kaxil** [comment]
> Hey @gerardo @dimberman @kaxil @ashb @mik-laj @nuclearpinguin
> 
> I think you all might be interested with this change to see/review/understand how kind implementation evolves and how Breeze will let people to work with Kubernetes Executor in development mode. Kind seems to be really nice for both CI and local development.
> 
> I have finally implemented the "best" approach with Kind I think. One that is much faster and seems to work really well for both - CI (much faster than previous approach with separate docker container) - it also helps with local development with kind cluster . Now I am testing all the ins/outs of it but it seems that it is really nice for local development as well (with regards to resources/time pulling images/using latest sources etc.). Together with the follow-up change of separating integration and slimming down Breeze: #7091 - it becomes really useful for local development.
> 
> Some of the notable properties:
> 
> * I switched to the latest kind 6.1 (it does not require separate exporting of the ".kube folder - it is done automatically)
> * I pass the docker socket from host to the container for "ENABLE_KIND_CLUSTER" build - this way I can use the same docker image that airflow-testing is run on - all the Kind containers are run in that docker image (and visible from the host via docker ps). This means a) no separate docker container is needed b) no need to pull image from apache/airflow - it's already there! I am checking if container is already created and skip creating it if so. You can also force cluster recreation by passing --recreate-kind cluster flag or stop
> * this also means that if I exit the Breeze/airflow-testing container - the cluster is still running there and I do not recreate the cluster when I re-enter the container.
> * However I still run "rebuild_airflow_image.sh" when entering the container. This way I am 100% sure I use latest sources (both locally and in CI). Plus I can later change the sources and run "rebuild_airflow_image.sh" without leaving the airflow-testing container! What's nice about it - it's rather fast as it only rebuilds latest layers. It allows to iterate semi-quickly while working on Kubernetes :). Still loading image to kind takes quite some time (6 minutes for me) so it is not perfect. I will improve it when I have prod image - it should be much faster to load rather than the CI one
> * last but not least - the .kube folder is mounted from the HOST (${AIRFLOW_SOURCES}/.kube) so you should be able to interact with the cluster from the host as well (You just need to update hostname to localhost). I might want to document it or maybe even automatically copy the configuration to include the configuration that will support kubectl/connecting to cluster-run webserver from the localhost directly. All ports are exposed already so it should be rather easy.

Nice work @potiuk

**Comment by potiuk** [comment]
I think I solved all the problems and hopefully it will go green now. There were a few problems with kind on docker mapped from host - namely CoreDNS was looping and crashed and it seems we have hard-coded (or possibly I could not find it) airflow:latest as the image that is used by the KubernetesExecutor to start workers (@dimberman - I would appreciate if you take a look at this change as well - you might be able to use the Kind cluster to add tests for KnativeExecutor that you might be ablet to run in CI).

**Comment by potiuk** [comment]
Yep. It's Green now. All test working -> I can finish now the change to separate out integration tests based on this. Pls. take a look

File: BREEZE.rst [hunk 7]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+Building Airflow Images and Loading them to Kubernetes cluster
+--------------------------------------------------------------
+
+This is done using ``./scripts/ci/in_container/kubernetes/docker/rebuild_airflow_image.sh`` script:
+
+1. Latest ``apache/airflow:master-pythonX.Y-ci`` images are rebuilt using latest sources.
+3. New kubernetes image based on the  ``apache/airflow:master-pythonX.Y-ci`` is build with
+   necessary scripts added to run in kubernetes. The image is tagged with
+   ``apache/airflow:master-pythonX.Y-ci-kubernetes`` tag.
+4. The image is loaded to the kind cluster using ``kind load`` command from kind.
```
Comment (potiuk) [inline BREEZE.rst]:
```suggestion
4. The image is loaded to the kind cluster using ``kind load`` command with ``airflow:latest`` tag.
```

File: BREEZE.rst [hunk 8]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+automatically create a Kind Kubernetes cluster in the same ``docker`` engine that the user uses to run Breeze
+Setting up the Kubernetes cluster takes some time, so the cluster continues running
+until the user stops the cluster explicitly with the  ``--stop-kind-cluster`` switch or until the user runs breeze with the ``--recreate-kind-cluster``
+switch (rather than ``--start-kind-cluster``).
+
+The cluster name follows the pattern ``airflow-python-X.Y.Z-vA.B.C`` where X.Y.Z is the Python version,
+and A.B.C is kubernetes version. This allows the user to have multiple clusters running at the same
+time for different python versions and different kubernetes versions.
+
+The Control Plane  is available from inside the docker image via <CLUSTER_NAME>-control-plane:6443
```
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
The Control Plane is available from inside the docker image via ``<CLUSTER_NAME>-control-plane:6443``
```

File: BREEZE.rst [hunk 9]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+host:port, the worker of the kind cluster is available at  <CLUSTER_NAME>-worker
+and webserver port for the worker is 30809.
+
+The Kubernetes Cluster is started but in oder to deploy airflow to Kubernetes cluster you need to:
```
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
The Kubernetes Cluster is started but in order to deploy airflow to Kubernetes cluster you need to:
```

File: BREEZE.rst [hunk 10]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+
+1. Build the image.
+2. Load it to Kubernetes cluster.
+3. Deploy airflow application.
+
+It can be done with single script: ``./scripts/ci/in_container/kubernetes/deploy_airflow_to_kubernetes.sh``
+
+You can however work separately on image in Kubernetes and the deploying the Airflow app in the cluster.
```
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
You can, however, work separately on the image in Kubernetes and deploying the Airflow app in the cluster.
```

File: BREEZE.rst [hunk 11]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+
+Building Airflow Images and Loading them to Kubernetes cluster
+--------------------------------------------------------------
+
+This is done using ``./scripts/ci/in_container/kubernetes/docker/rebuild_airflow_image.sh`` script:
+
+1. Latest ``apache/airflow:master-pythonX.Y-ci`` images are rebuilt using latest sources.
+3. New kubernetes image based on the  ``apache/airflow:master-pythonX.Y-ci`` is build with
```
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
2. New Kubernetes image based on the  ``apache/airflow:master-pythonX.Y-ci`` is built with
```

File: BREEZE.rst [hunk 12]
```diff
@@ -528,6 +509,65 @@ As soon as you enter the Breeze environment, you can run Airflow unit tests via
...
+   necessary scripts added to run in kubernetes. The image is tagged with
+   ``apache/airflow:master-pythonX.Y-ci-kubernetes`` tag.
+4. The image is loaded to the kind cluster using ``kind load`` command
```
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
3. The image is loaded to the kind cluster using ``kind load`` command
```

File: breeze [hunk 13]
```diff
@@ -82,15 +82,30 @@ TEST_TARGET=""
...
+export AIRFLOW_CONTAINER_DEPS="true"
+
+# Do not start Kubernetes cluster by default
+export ENABLE_KIND_CLUSTER="false"
+
+# Do not recreate Kubernetes cluster by default
+export RECREATE_KIND_CLUSTER="false"
+
+# Do not stop Kubernetes cluster by default
+export SDTOP_KIND_CLUSTER="false"
```
Comment (kaxil) [inline breeze]:
```suggestion
export STOP_KIND_CLUSTER="false"
```

File: scripts/ci/in_container/entrypoint_ci.sh [hunk 14]
```diff
@@ -37,15 +37,16 @@ TRAVIS=${TRAVIS:=}
 
 AIRFLOW_SOURCES=$(cd "${MY_DIR}/../../.." || exit 1; pwd)
 
-PYTHON_VERSION=${PYTHON_VERSION:=3.6}
-ENV=${ENV:=docker}
+PYTHON_VERSION=${PYTHON_VERSION:=3.5}
```
Comment (kaxil) [inline scripts/ci/in_container/entrypoint_ci.sh]:
```suggestion
PYTHON_VERSION=${PYTHON_VERSION:=3.6}
```

File: scripts/ci/local_ci_enter_environment.sh [hunk 15]
```diff
@@ -21,7 +21,7 @@ MY_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 export RUN_TESTS="false"
 export MOUNT_LOCAL_SOURCES="true"
-export PYTHON_VERSION=${PYTHON_VERSION:="3.6"}
+export PYTHON_VERSION=${PYTHON_VERSION:="3.5"}
```
Comment (kaxil) [inline scripts/ci/local_ci_enter_environment.sh]:
```suggestion
export PYTHON_VERSION=${PYTHON_VERSION:="3.6"}
```

**Comment by potiuk** [comment]
@nuclearpinguin @kaxil @mik-laj - one last final review and I'd love to merge it

**Review by kaxil — APPROVED** [review]

**Pull request merged**